quickwit/quickwit-storage/tests/google_cloud_storage.rs

```diff
@@ -16,29 +16,90 @@
-
-    let mut object_storage = new_emulated_google_cloud_storage(&Uri::from_str(
-        "gs://sample-bucket/integration-tests/test-azure-compatible-storage",
-    )?)?;
-    quickwit_storage::storage_test_single_part_upload(&mut object_storage)
-        .await
-        .context("test single-part upload failed")?;
-
-    quickwit_storage::storage_test_multi_part_upload(&mut object_storage)
-        .await
-        .context("test multipart upload failed")?;
-    Ok(())
+    use quickwit_storage::test_config_helpers::{
+        DummyTokenLoader, LOCAL_GCP_EMULATOR_ENDPOINT, new_emulated_google_cloud_storage,
+    };
+    use reqsign::GoogleTokenLoad;
+
+    pub async fn sign_gcs_request(req: &mut reqwest::Request) -> anyhow::Result<()> {
+        let client = reqwest::Client::new();
+        let token = DummyTokenLoader
+            .load(client.clone())
+            .await?
+            .ok_or_else(|| anyhow::anyhow!("Failed to obtain authentication token"))?;
+
+        let signer = reqsign::GoogleSigner::new("storage");
+        signer.sign(req, &token)?;
+
+        Ok(())
+    }
+
+    async fn create_gcs_bucket(bucket_name: &str) -> anyhow::Result<()> {
+        let client = reqwest::Client::new();
+        let url = format!("{LOCAL_GCP_EMULATOR_ENDPOINT}/storage/v1/b");
+        let mut request = client
+            .post(url)
+            .body(serde_json::to_vec(&serde_json::json!({
+                "name": bucket_name,
+            }))?)
+            .header(reqwest::header::CONTENT_TYPE, "application/json")
+            .build()?;
+
+        sign_gcs_request(&mut request).await?;
+
+        let response = client.execute(request).await?;
+
+        if !response.status().is_success() {
+            let error_text = response.text().await?;
+            anyhow::bail!("Failed to create bucket: {}", error_text);
+        };
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn google_cloud_storage_test_suite() -> anyhow::Result<()> {
+        setup_logging_for_tests();
+
+        let bucket_name = append_random_suffix("sample-bucket").to_lowercase();
+        create_gcs_bucket(bucket_name.as_str())
+            .await
+            .context("Failed to create test GCS bucket")?;
+
+        let mut object_storage =
+            new_emulated_google_cloud_storage(&Uri::from_str(&format!("gs://{bucket_name}"))?)?;
+
+        quickwit_storage::storage_test_suite(&mut object_storage).await?;
+
+        let mut object_storage = new_emulated_google_cloud_storage(&Uri::from_str(&format!(
+            "gs://{bucket_name}/integration-tests/test-gcs-storage"
+        ))?)?;
+
+        quickwit_storage::storage_test_single_part_upload(&mut object_storage)
+            .await
+            .context("test single-part upload failed")?;
+
+        // TODO: Uncomment storage_test_multi_part_upload when the XML API is
+        // supported in the emulated GCS server
+        // (https://github.com/fsouza/fake-gcs-server/pull/1164)
+
```

Comment:
were you able to confirm this test would pass on a real GCS storage?

Reply:
yes, and I checked in the GCS audit logs that it was indeed a multipart upload (it didn't show the part size but there were 3 parts which matches 15MB / 5MB)